Source/Core/DolphinQt2/Config/GeckoCodeWidget.cpp

```diff
@@ -0,0 +1,205 @@
+  m_code_view = new QListWidget;
+  m_download_codes = new QPushButton(tr("Download Codes (WiiRD Database)"));
+
+  m_code_description->setReadOnly(true);
+
+  m_download_codes->setEnabled(!m_game_id.empty());
+
+  layout->addWidget(m_warning);
+  layout->addWidget(m_code_list, 3);
+
+  auto* info_layout = new QFormLayout;
+
+  info_layout->addRow(tr("Name:"), m_name_label);
+  info_layout->addRow(tr("Creator:"), m_creator_label);
+
+  for (QLabel* label : {m_name_label, m_creator_label})
+  {
+    label->setTextInteractionFlags(Qt::TextSelectableByMouse);
+    label->setCursor(Qt::IBeamCursor);
+  }
+
+  layout->addWidget(new QLabel(tr("Description:")));
+  layout->addLayout(info_layout, 0);
+  layout->addWidget(m_code_description, 0);
+  layout->addWidget(m_code_view, 0);
+  layout->addWidget(m_download_codes, 0, Qt::AlignRight);
+
+  m_download_codes->setSizePolicy(QSizePolicy::Fixed, QSizePolicy::Fixed);
+
+  setLayout(layout);
+}
+
+void GeckoCodeWidget::ConnectWidgets()
+{
+  connect(m_code_list, &QListWidget::itemSelectionChanged, this,
+          &GeckoCodeWidget::OnSelectionChanged);
+  connect(m_code_list, &QListWidget::itemChanged, this, &GeckoCodeWidget::OnItemChanged);
+
+  connect(m_download_codes, &QPushButton::pressed, this, &GeckoCodeWidget::DownloadCodes);
+
+  connect(m_warning, &CheatWarningWidget::OpenCheatEnableSettings, this,
+          &GeckoCodeWidget::OpenGeneralSettings);
+
+  connect(this, &GeckoCodeWidget::EmulationStarted, m_warning,
+          &CheatWarningWidget::EmulationStarted);
+  connect(this, &GeckoCodeWidget::EmulationStopped, m_warning,
+          &CheatWarningWidget::EmulationStopped);
+}
+
+void GeckoCodeWidget::OnSelectionChanged()
+{
+  auto items = m_code_list->selectedItems();
+
+  if (items.empty())
+    return;
+
+  auto selected = items[0];
+
+  const auto& code = m_gecko_codes[m_code_list->row(selected)];
+
+  QString notes;
+
+  for (const std::string& line : code.notes)
+    notes.append(QString::fromStdString(line).append(QLatin1Char('\n')));
+
+  m_name_label->setText(QString::fromStdString(code.name));
+  m_creator_label->setText(QString::fromStdString(code.creator));
+
+  QFont monospace(QFontDatabase::systemFont(QFontDatabase::FixedFont).family());
+
+  m_code_description->setHtml(tr("<pre>%2</pre>").arg(notes.isEmpty() ? tr("N/A") : notes));
+
+  m_code_description->setFont(monospace);
+
+  m_code_view->clear();
+
+  m_code_view->setFont(monospace);
+
+  for (const auto& c : code.codes)
+    m_code_view->addItem(QStringLiteral("%1  %2")
+                             .arg(c.address, 8, 16, QLatin1Char('0'))
+                             .arg(c.data, 8, 16, QLatin1Char('0')));
+}
+
+void GeckoCodeWidget::OnItemChanged(QListWidgetItem* item)
+{
+  m_gecko_codes[m_code_list->row(item)].enabled = (item->checkState() == Qt::Checked);
+
+  SaveCodes();
+}
+
+void GeckoCodeWidget::SaveCodes()
+{
+  auto game_ini_local = SConfig::GetInstance().LoadLocalGameIni(m_game_id, m_game_revision);
```

Comment:
This will load from all possible user gameinis, and then write all of that information into either the revision-specific ini or the full-gameid-specific ini. Instead, it should figure out which ini it will write to, and only read that one in (i.e., we want to merge the new gecko code configuration into whatever's there, not possibly other copy other settings around).